verify.js

```diff
@@ -85,97 +62,135 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
-
-    if (invalid_issuer) {
-      return done(new JsonWebTokenError('jwt issuer invalid. expected: ' + options.issuer));
-    }
-  }
-
-  if (options.subject) {
-    if (payload.sub !== options.subject) {
-      return done(new JsonWebTokenError('jwt subject invalid. expected: ' + options.subject));
-    }
-  }
-
-  if (options.jwtid) {
-    if (payload.jti !== options.jwtid) {
-      return done(new JsonWebTokenError('jwt jwtid invalid. expected: ' + options.jwtid));
-    }
-  }
-
-  if (options.maxAge) {
-    if (typeof payload.iat !== 'number') {
-      return done(new JsonWebTokenError('iat required when maxAge is specified'));
-    }
-
-    var maxAgeTimestamp = timespan(options.maxAge, payload.iat);
-    if (typeof maxAgeTimestamp === 'undefined') {
-      return done(new JsonWebTokenError('"maxAge" should be a number of seconds or string representing a timespan eg: "1d", "20h", 60'));
-    }
-    if (clockTimestamp >= maxAgeTimestamp + (options.clockTolerance || 0)) {
-      return done(new TokenExpiredError('maxAge exceeded', new Date(maxAgeTimestamp * 1000)));
-    }
-  }
-
-  return done(null, payload);
+  var getSecret;
+
+  if(typeof secretOrPublicKey === 'function') {
+      if(!callback) {
+          return done(new JsonWebTokenError('verify must be called asynchronous if secret or public key is provided as a callback'));
+      }
+
+      getSecret = secretOrPublicKey;
+  }
+  else {
+      getSecret = function(header, secretCallback) {
+          return secretCallback(null, secretOrPublicKey);
+      };
+  }
+
+  return getSecret(header, function(err, secretOrPublicKey) {
+      if(err) {
+          return done(new JsonWebTokenError('error in secret or public key callback: ' + err.message));
+      }
+
+      var hasSignature = parts[2].trim() !== '';
+
+      if (!hasSignature && secretOrPublicKey){
+        return done(new JsonWebTokenError('jwt signature is required'));
+      }
+
+      if (hasSignature && !secretOrPublicKey) {
+        return done(new JsonWebTokenError('secret or public key must be provided'));
+      }
+
+      if (!hasSignature && !options.algorithms) {
+          options.algorithms = ['none'];
+      }
+
+      if (!options.algorithms) {
+          options.algorithms = ~secretOrPublicKey.toString().indexOf('BEGIN CERTIFICATE') ||
+          ~secretOrPublicKey.toString().indexOf('BEGIN PUBLIC KEY') ?
+              ['RS256', 'RS384', 'RS512', 'ES256', 'ES384', 'ES512'] :
+              ~secretOrPublicKey.toString().indexOf('BEGIN RSA PUBLIC KEY') ?
+                  ['RS256', 'RS384', 'RS512'] :
+                  ['HS256', 'HS384', 'HS512'];
+
+      }
+
+      if (!~options.algorithms.indexOf(decodedToken.header.alg)) {
+          return done(new JsonWebTokenError('invalid algorithm'));
+      }
+
+      var valid;
+
+      try {
+          valid = jws.verify(jwtString, decodedToken.header.alg, secretOrPublicKey);
+      } catch (e) {
+          return done(e);
+      }
+
+      if (!valid)
```

Comment:
please be consistent on the braces usage. A few lines above this one, another one liner if clause uses them.

Reply:
Noticed this as well, but because it was already on master this way. I was planning to add eslint and fixing it one go.